api/envoy/config/core/v3/http_service.proto

```diff
@@ -0,0 +1,48 @@
+syntax = "proto3";
+
+package envoy.config.core.v3;
+
+import "envoy/config/core/v3/base.proto";
+
+import "google/protobuf/duration.proto";
+
+import "udpa/annotations/status.proto";
+import "validate/validate.proto";
+
+option java_package = "io.envoyproxy.envoy.config.core.v3";
+option java_outer_classname = "HttpServiceProto";
+option java_multiple_files = true;
+option go_package = "github.com/envoyproxy/go-control-plane/envoy/config/core/v3;corev3";
+option (udpa.annotations.file_status).package_version_status = ACTIVE;
+
+// [#protodoc-title: HTTP services]
+
+// HTTP service configuration.
+// [#next-free-field: 6]
+message HttpService {
```

Comment:
Did you model this on ext_authz `HttpService`? I notice there are a few subtle differences, e.g. use of `HttpUri` there.

Reply:
No, I didn't model based on that. I'm not sure, but I don't see a point in using the `HttpUri` type:

It forces defining a URI in the config, which won't be used in the OTLP exporter. In this PR, the exporter obtains an HTTP client from the thread local cluster which already has the base url configured. This was something that I noticed in @AlexanderEllis original PR and commented about it https://github.com/envoyproxy/envoy/pull/28454#issuecomment-1650106711. In the old PR, I could put whatever I wanted in `uri:` and it didn't matter. 

I looked, the `ext_authz` seems to be doing the same https://github.com/envoyproxy/envoy/blob/main/source/extensions/filters/common/ext_authz/ext_authz_http_impl.cc#L254 and I also don't see any usage of the `Uri` there. 

And by looking at this example https://github.com/envoyproxy/envoy/blob/main/examples/ext_authz/config/http-service.yaml#L32, unless I'm missing something, I don't see where `ext_authz` is coming from (used in `uri`), which confirms that it's just there to satisfy the field in `HttpUri` being required.

Reply:
Interesting… it took me a while to dig through the history of `HttpUri`, and found [this comment thread](https://github.com/envoyproxy/data-plane-api/pull/530#discussion_r173567937) interesting in particular while trying to understand what was the motivation behind `HttpUri.uri`.
Seems it was introduced to set the path, scheme and other parts of the request, while `HttpUri.cluster` sets the used Envoy cluster.

The 2 places I saw it is used as intended seem to be:
1. JWKS fetcher
2. Remote data resource

While the rest seem to be using just the `cluster` part.

I agree that ext-authz seems to be using it just for the `cluster`, and disregards the `uri`, but maybe it should have used it as intended (maybe in the future?).
I think it will be better to use `HttpUri` as is and probably try to reuse some of the code from [jwks fetcher](https://github.com/envoyproxy/envoy/blob/94a69e5dc11ecc76e227ed5e366fa15e857aff9e/source/extensions/filters/http/common/jwks_fetcher.cc#L47-L75) for this use-case.

Reply:
@adisuissa I changed it to use `HttpUri` following the jwks fetcher example you gave [3be83f8](https://github.com/envoyproxy/envoy/pull/29207/commits/3be83f8c10c076c7107c04ec84183c3b5c0be372). Looks a bit simpler now :)

Reply:
@htuch @adisuissa ,
Noticed this when working on http design for ext_proc. 

I am wondering why we add this `HttpService` in core proto? More specifically, `http_uri` is already a core proto that is generic and being used by several components, but  `request_headers_to_add` here seems not be a generic field
Instead, I can see another approach here :  Wrap `http_uri` (core config field) and extension-specific fields into extension's own proto file , which is similar to [ext_authz](https://github.com/envoyproxy/envoy/blob/4991c0fcc820671a0e6b1421da5f5de423022380/api/envoy/extensions/filters/http/ext_authz/v3/ext_authz.proto#L277). 
By doing so, every extension can share the common `http_uri` field and add its own fields on demand.

Reply:
Why isn't `request_headers_to_add` a generic thing for an HTTP service? Seems a standard thing you might want to do when invoking an HTTP service, e.g. to set an auth token.

Reply:
From HTTP service perspective, yes `request_headers_to_add` is generic thing.
However, what I meant is from user and  practical perspective, this won't necessarily be achieved by this field in base proto here. More specifically:

-  oauth [add token](https://github.com/envoyproxy/envoy/blob/1ef48e62bfff7be0403d6609eb9f57e17907bcc2/source/extensions/filters/http/oauth2/oauth_client.cc#L85C48-L85C57) based on client_id and secret from [its oauth.proto](https://github.com/envoyproxy/envoy/blob/1ef48e62bfff7be0403d6609eb9f57e17907bcc2/api/envoy/extensions/filters/http/oauth2/v3/oauth.proto#L55-L58); ext_authz add the headers from its [attribute_context.proto](https://github.com/envoyproxy/envoy/blob/1ef48e62bfff7be0403d6609eb9f57e17907bcc2/api/envoy/service/auth/v3/attribute_context.proto#L103); 
- For ext_proc, I would think headers is added from its [processor.proto](https://github.com/envoyproxy/envoy/blob/1ef48e62bfff7be0403d6609eb9f57e17907bcc2/api/envoy/service/ext_proc/v3/external_processor.proto#L198) rather than using this field here.

In short, I don't have strong opinion here but I am not sure about practical value of this `request_headers_to_add ` field, as it can be achieved by extension specific config easily. It is probably good for simple add-header needs.

Reply:
I feel those make sense for "static" header values. If there's a use case for cases where a header value is not know in advance, like the client_id example, those can still be added "on the fly". I think one does not negate the other. 

The [GrpcService](https://github.com/envoyproxy/envoy/blob/main/api/envoy/config/core/v3/grpc_service.proto#L302) also has a headers property, so the goal was to be consistent.

Reply:
Yeah, I think for some simple use cases, like service specific static auth tokens, this is still reasonable.

Reply:
Yep. I agree with its value for simple use case.

Thanks all for discussion.